### ISSUE TITLE
optimization: let out.x have the same shape as x0

### ISSUE THREAD

**Issue by nschloe** [scipy/scipy#13869]
This is a suggestion for a backwards-incompatible change, so that's probably for SciPy 2.0.

All scipy optimization methods evaluate the objective function at points of shape `(n,)`, even if the initial guess is from a different domain. Here's an optimzation problem that converges to the solution 0 (pointwise):
```python
import numpy as np
from scipy.optimize import minimize


def f(x):
    print(f"{x.shape = }")
    return np.sum(x ** 2)

x0 = # something

out = minimize(f, x0)

print(f"{out.x.shape = }")
```
 * For `x0 = [0.7]` you get `(1,)` :heavy_check_mark: 
 * For `x0 = [0.7, 0.2]` you get `(2,)` :heavy_check_mark: 
 * For `x0 = 0.7` you get `(1,)` (not `()`) :x: 
 * For `x0 = [[0.7, 0.2]]` you get `(2,)` (not `(1, 2)`) :x: 
 * For `x0 = [[0.7, 0.2], [0.3, -4.1]]` you get `(4,)` (not `(2, 2)`) :x: 

There is no mathematical reason for the reshape. The domain space of optimization methods must only be an inner-product space (i.e., `+` and `-` and scalar multiplication must make sense, plus you have to have a dot product for some methods). I would like to propose to

 * where possible, allow for inputs of any shape without reshape (e.g., Nelder-Mead)
 * where dot products are needed, reject input arrays of `len(x0.shape) > 1` (because `np.dot(a, b)` doesn't return a scalar there) _OR_ allow providing a custom dot product (this could be added later)

**Comment by tupui** [comment]
So in case of a scalar, you would like to return a scalar? I am not really in favour of the change. It's better to be consistent with the return types. We should actually do the opposite and force particular shapes at the input. Just like scikit-learn is doing.

**Comment by nschloe** [comment]
> Just like scikit-learn is doing.

Interesting. Can you point me to what they are doing?

> We should actually do the opposite and force particular shapes at the input.

The domain of `x` can be anything (n dimensions), so we can't force anything at the input anyway. The solution of the optimization should be in the domain too, of course.

**Comment by tupui** [comment]
> > Just like scikit-learn is doing.
> 
> Interesting. Can you point me to what they are doing?

[Here](https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.GaussianProcessRegressor.html#sklearn.gaussian_process.GaussianProcessRegressor) for instance. But they are doing this everywhere. You must feed an array of shape (n_samples, n_features). In the past they had a message inviting you to reshape and now they just raise if you don't do this.

> > We should actually do the opposite and force particular shapes at the input.
> 
> The domain of `x` can be anything (n dimensions), so we can't force anything at the input anyway. The solution of the optimization should be in the domain too, of course.

Sure, what I mean is to force a certain structure of the array and not allow different types of different shapes structure to be returned.

**Comment by nschloe** [comment]
Well, I don't know the context of scikit-learn too well, but I can say that in optimization methods the structure of the domain is used nowhere. The only important thing is that it's an inner product space (in the sense that the operations +, -, scalar multiplication, and dot product make sense). Some methods don't even require the dot product. In principle, the shape of the `x0` could be anything and all operations should work.

**Comment by tupui** [comment]
I am not sure I understand what you mean. Could you post some pseudo-code to explain what you propose and why it's important vs what we currently have? Also for breaking changes I suggest you use the mail list for that.

**Comment by nschloe** [comment]
Let me adapt the original post.

**Comment by nschloe** [comment]
@tupui I edited the original post and incorporated your suggestion to reject certain inputs (where necessary). I think the proposal makes more sense now, so thanks for the input.

**Comment by tupui** [comment]
Ok now I get it better, thanks for updating.

I think it's just not design from the start with that in mind. `minimize` assumes scalar inputs. The doc states: `Minimization of scalar function of one or more variables.` also look at the definition of `x0` `Array of real elements of size (n,), where ‘n’ is the number of independent variables.`.

So what you propose would require to implement some optimizer which handle vectorial inputs.

**Comment by nschloe** [comment]
> So what you propose would require to implement some optimizer which handle vectorial inputs.

All algorithms that don't require an inner product (Nelder-Mead and Powell, for example), can already do that with little or no change.

**Comment by tupui** [comment]
Yes but we would want to stay consistent along all optimizers. So it would need to be thought through in terms of API.

I would suggest you send an email to the dev mailing list with your proposal. Also state if you are actually proposing to do the work.

**Comment by nschloe** [comment]
> Yes but we would want to stay consistent along all optimizers.

Yeah, that's in the proposal, right? "Don't reshape the domain space, and reject input x0 for which there's no meaningful dot-product (if it's required in the algo)."

> I would suggest you send an email to the dev mailing list with your proposal. Also state if you are actually proposing to do the work.

Will do.

**Comment by andyfaff** [comment]
`minimize` is for functions R^m -> R^1. As such I  think it makes sense that `x` has shape (m,). It doesn't make sense for `x` to be >1D because Hessians and gradients would have too many dimensions, and doesn't agree with R^m.

Whilst it's a pain if objective functions return more than a scalar it's not clear to me (from a pragmatic POV) what benefits are gained from x0 having the same shape as out.x.

**Comment by nschloe** [comment]
Thanks Andy for taking a look!

> minimize is for functions R^m -> R^1. As such I think it makes sense that x has shape (m,). 

Right, that's a necessary restriction for some methods (namely those which require an inner product, e.g., CG, BFGS). The funny thing is that scipy will still accept other `x0` (e.g., matrices) and will simply reshape them into "column-vectors" of shape `(n, 1)`. I'd call the fact that this happens silently.. questionable. But realistically, that's probably not something we'll ever tackle. Other methods don't have that requirement (e.g., Nelder-Mead, Powell ) since they don't use an inner product. (They just scale, form averages etc.) Input points are still getting reshaped.

> Whilst it's a pain if objective functions return more than a scalar

Well, optimization doesn't even make sense then. I think current scipy correctly raises and exception in this case. Anyway, this is a different issue.

> what benefits are gained from x0 having the same shape as out.x.

For speed, there will be no difference. I'm only suggesting to remove some unnecessary reshapes. I think it's nothing more than natural if the solution `out.x` and all evaluation points `x` are in the same space as `x0`. That's the mathematical reality and could very well be reflected in code. The code would be cleaner, not only on scipy's side but also on the user side. For simple 1D minimization, I've always wondered why you need the `[0]` at the end:
```python
from scipy.optimize import minimize

def f(x):
    return x ** 2 - 2

x0 = 1.5
out = minimize(f, x0)
sol = out.x[0]
```
For higher-dimensional problems, it saves a reshape at the end,
```python
sol = np.,reshape(out.x, x0.shape)
```
This is actually my main motivation for suggesting the change.

**Comment by newville** [comment]
I agree with @andfaff, and a -1 from me.  In short, `x` should be a strictly 1-dimensional array.  

In more detail: Variables in a minimization problem should be a *set*.  Each variable has a scalar (and continuous) value and each must be unique.  You might consider that they have other attributes (name, bounds, standard error) and so be mapped Objects instead of plain floats -- we do this in lmfit, but it would be a large change to do that for scipy.optimize.   There might be benefits, but there would be costs too. 

In principle the order of the variables does not matter (and any fit can reorder the variables and get the same result).  But it is completely reasonable to expect that an order will be maintained by a fitting algorithm.  In practice and historically (as the algorithms mostly originated in Fortran or perhaps early C implementation), it makes a lot of sense for the algorithm to combine these into a Vector or Array of floats that happens to map exactly to a 1D numpy array, and then it very much has a reliable order that cannot be changed and the only way to access individual variables is by indexing the array.

Importantly, a grouping of variables is not multi-dimensional: that second axis never has meaning, and there is never meaning to transposing or reducing the dimensionality in any way other than "flatten to strictly 1-dimension".   

There is always ambiguity about whether arrays with 1-element are better interpreted as scalars.  For the specific problem here, please maintain that these, and all groupings of variables in scipy.optimize are 1-dimensional numpy arrays.
